### PR TITLE
Fix: check for objects or arrays in array-bracket-spacing (fixes #4083)

### LINKED ISSUE #4083: array-bracket-spacing false-positive when accessing array index in array.
I have the following in my .eslintrc file:
```json
"array-bracket-spacing" : [ 1, "always", {
  "objectsInArrays": false,
  "arraysInArrays": false
}]
```
...and the following code:
```js
var theBestApples = [ apples[i], apples[j] ]
```
... and the following ```array-bracket-spacing``` warning from eslint: ```There should be no space before ']' at line...```

My guess is that eslint is seeing ```[j]``` and incorrectly recognizing it as an array literal rather than reference of index ```j``` on the ```apples``` array. The warning only gets triggered when the array reference is the last element in the array. Eliminating the space before the final ```]``` removes the warning.

I am using eslint version ```1.1.0```.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/27259027-array-bracket-spacing-false-positive-when-accessing-array-index-in-array?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by alberto**

### CODE REVIEW

File: lib/rules/array-bracket-spacing.js [hunk 1]
```diff
@@ -123,17 +123,19 @@ module.exports = function(context) {
...
             penultimate = context.getLastToken(node, 1),
-            last = context.getLastToken(node);
+            last = context.getLastToken(node),
+            firstElement = node.elements[0],
+            lastElement = node.elements[node.elements.length - 1];
 
         var openingBracketMustBeSpaced =
-            options.objectsInArraysException && second.value === "{" ||
-            options.arraysInArraysException && second.value === "[" ||
+            options.objectsInArraysException && (firstElement.type === "ObjectExpression" || firstElement.type === "ObjectPattern") ||
```
Comment (nzakas) [inline lib/rules/array-bracket-spacing.js]: Since you're duplicating the type check in multiple places, can you split out into a function?

Comment (alberto) [inline lib/rules/array-bracket-spacing.js]: Do you mean to extract a check for each of the types (object and array) or a `mustElementBeSpaced(element)` function to extract the whole check?

Comment (nzakas) [inline lib/rules/array-bracket-spacing.js]: Each types.

Comment (alberto) [inline lib/rules/array-bracket-spacing.js]: Done

**Comment by nzakas** [comment]
Nice! LGTM

**Pull request merged**